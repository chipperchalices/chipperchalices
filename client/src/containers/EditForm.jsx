import React, {Component} from 'react';
import { connect } from 'react-redux';
import axios from 'axios';
import moment from 'moment';
import {DatePicker, TimePicker} from 'material-ui';
import { Button, Header, Image, Modal, Icon, Form, Dropdown, TextArea } from 'semantic-ui-react';

import { fromStatus } from '../helpers/status.js';
import { moveCard } from '../actions/index.js';

class EditForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      modalOpen: false,
      title: '',
      company: '',
      date: moment().toDate(),
      // statusDate:
      status: '',
      notes: '',
      url: '',
      value: 0,
      change: false,
      recruiter_name: '',
      recruiter_email: ''
    };

    this.handleOpen = this.handleOpen.bind(this);
    this.handleClose = this.handleClose.bind(this);
    this.handleTitle = this.handleTitle.bind(this);
    this.handleCompany = this.handleCompany.bind(this);
    this.handleDate = this.handleDate.bind(this);
    this.handleStatus = this.handleStatus.bind(this);
    this.saveJob = this.saveJob.bind(this);
    this.handleNotes = this.handleNotes.bind(this);
    this.handleURL = this.handleURL.bind(this);
    this.handleRecruiter = this.handleRecruiter.bind(this);
    this.handleEmail = this.handleEmail.bind(this);
  }

  //TO DO: Refactor State into Singular Object
  componentDidMount() {
    let dat = this.props.job;
    this.setState({
      modalOpen: false,
      id: dat.id,
      title: dat.position,
      company: dat.company.name,
      company_id: dat.company.id,
      date: moment(dat.updated_at).toDate(),
      //date: dat.updated_at,
      status: dat.current_status,
      notes: dat.notes,
      url: dat.position_url,
      value: 0,
      recruiter_name: dat.recruiter_name,
      recruiter_email: dat.recruiter_email
    });
  }

  //Move to Action/Reducer in Redux
  //TO DO: refactor into less functions

  handleOpen() {
    this.setState({ modalOpen: true });
  }

  handleClose() {
    this.setState({ modalOpen: false });
  }

  handleTitle (e) {
    this.setState({
      change: true,
      title: e.target.value
    });
  }

  handleCompany (e) {
    this.setState({
      change: true,
      company: e.target.value
    });
  }

  handleRecruiter (e) {
    this.setState({
      recruiter_name: e.target.value
    });
  }

  handleEmail (e) {
    this.setState({
      recruiter_email: e.target.value
    });
  }

  handleNotes (e) {
    this.setState({
      change: true,
      notes: e.target.value
    });
  }

  handleURL (e) {
    this.setState({
      change: true,
      url: e.target.value
    });
  }

  handleStatus (event, index, value) {
    this.setState({
      change: true,
      status: event.target.textContent,
      value: value,
    });
  }

  handleDate (e, date) {
    this.setState({
      change: true,
      date: moment(date).toDate()
    });
  }

  saveJob (e) {
    //e.preventDefault();
    const { x, job, moveCard } = this.props;

    this.handleClose();

    //Is an action instead of a CB
    var form = this;
    // Add more to object later
    var content = {
      id: this.state.id,
      job: {
        title: this.state.title,
        company: this.state.company,
        notes: this.state.notes,
        url: this.state.url,
        recruiter_name: this.state.recruiter_name,
        recruiter_email: this.state.recruiter_email
      },
      status: {
        date: this.state.date,
        status: this.state.status
      }
    };

    let lastStatus = fromStatus(job.current_status);
    let nextStatus = fromStatus(this.state.status);

    moveCard(content, lastStatus, nextStatus, x);
  }

  render() {

    // console.log('after componentDidMount daate', this.state.date)
    // const { job } = this.props;
    // let slicedDate = job.updated_at.slice(0, -14)
    // console.log('date', job.updated_at, 'sliced', slicedDate)

    // const statusDate = moment(job.statusDate).format('YYYY-MM-DD');

    const options = [
      { key: 'Interested', text: 'Interested', value: 'Interested' },
      { key: 'Applied', text: 'Applied', value: 'Applied' },
      { key: 'Interview Scheduled', text: 'Interview Scheduled', value: 'Interview Scheduled' },
      { key: 'Interviewed', text: 'Interviewed', value: 'Interviewed' },
      { key: 'Applied', text: 'Applied', value: 'Applied' },
      { key: 'No response', text: 'No response', value: 'No response' },
      { key: 'Rejected', text: 'Rejected', value: 'Rejected' },
      { key: 'Offer made', text: 'Offer made', value: 'Offer made' },
      { key: 'Archived', text: 'Archived', value: 'Archived' },
    ];

    return (
      <Modal trigger={<Button size='mini' floated='right' color='white' circular icon='write' onClick={this.handleOpen} />}
        open={this.state.modalOpen}
        onClose={this.handleClose}
        size='large'>
        <Modal.Header>Edit Job</Modal.Header>
        <Modal.Content image>
          <Image wrapped size='medium' src={this.props.job.company.logo_url} />
          <Form>
            <Form.Group widths='equal'>
              <Form.Field>
                <label>Position</label>
                <input defaultValue={this.state.title} onChange={this.handleTitle}/>
              </Form.Field>
              <Form.Field>
                <label>Company</label>
                <input defaultValue={this.state.company} onChange={this.handleCompany}/>
              </Form.Field>
              <Form.Field>
                <label>Application Link</label>
                <input defaultValue={this.state.url} onChange={this.handleURL}/>
              </Form.Field>
            </Form.Group>
            <Form.Group widths='equal'>
              <Form.Field>
                <label>Recruiter Name</label>
                <input defaultValue={this.state.recruiter_name} onChange={this.handleRecruiter}/>
              </Form.Field>
              <Form.Field>
                <label>Recruiter Email</label>
                <input defaultValue={this.state.recruiter_email} onChange={this.handleEmail}/>
              </Form.Field>
            </Form.Group>
            <Form.Group>
              <Form.Field>
                <Button.Group color='teal'>
                  <Button>Status</Button>
                  <Dropdown onChange={this.handleStatus} options={options} floating button />
                </Button.Group>
              </Form.Field>
              <Form.Field>
                <DatePicker mode="landscape" onChange={this.handleDate} defaultValue={this.state.date} value={this.state.date}/>
              </Form.Field>
            </Form.Group>
            <Form.Field>
              <label>Notes</label>
              <TextArea defaultValue={this.state.notes} onChange={this.handleNotes} />
            </Form.Field>
          </Form>
        </Modal.Content>
        <Modal.Actions>
          <Button color='teal' onClick={this.saveJob}>
            <Icon name='checkmark' /> Update
          </Button>
        </Modal.Actions>
      </Modal>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    moveCard: (card, lastStatus, nextStatus, lastX) => dispatch(moveCard(card, lastStatus, nextStatus, lastX))
  };
};

export default connect(null, mapDispatchToProps)(EditForm);
