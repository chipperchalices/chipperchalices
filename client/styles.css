#newApp {
	body {
	    background-color: lightblue;
	}
}

#stage2 {
	body {
	    background-color: lightblue;
	}
}

#stage3 {
	body {
	    background-color: lightblue;
	}
}

#stage4 {
	body {
	    background-color: lightblue;
	}
}

#stage5 {
	body {
	    background-color: lightblue;
	}
}

#stage6 {
	body {
	    background-color: lightblue;
	}
}


