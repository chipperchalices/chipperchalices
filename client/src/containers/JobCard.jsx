import React, { Component } from 'react';
import { DragSource } from 'react-dnd';
import { Button, Icon, Card, Image, Grid, Modal, Segment, Label } from 'semantic-ui-react';
import ExpandedForm from './ExpandedForm.jsx';
import EditForm from './EditForm.jsx';
const moment = require('moment');
const now = moment();

class JobCard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      age: moment().diff('days', moment(this.props.job.statusDate)),
      status: this.props.job.current_status
    };


    this.updateJob = this.updateJob.bind(this);
  }

  updateJob(data) {
    console.log(data);
  }

  render() {
    const { x, job, id } = this.props;
    const start = moment(job.statusDate).format('YYYY-MM-DD');

    var age = now.diff(moment(start), 'days');

    var statusChange, calendar;

    if (age <= 1) {
      statusChange = <Label circular color='green' inverted><Icon name='time' /> 1 day</Label>;
    } else if (age === 2) {
      statusChange = <Label circular color='green' inverted><Icon name='time' /> 2 days</Label>;
    } else if (age === 3) {
      statusChange = <Label circular color='green' inverted><Icon name='time' /> 3 days</Label>;
    } else if (age === 4) {
      statusChange = <Label circular color='olive' inverted><Icon name='time' /> 4 days</Label>;
    } else if (age === 5) {
      statusChange = <Label circular color='olive' inverted><Icon name='time' /> 5 days</Label>;
    } else if (age === 6) {
      statusChange = <Label circular color='olive' inverted><Icon name='time' /> 6 days</Label>;
    } else if (age === 7) {
      statusChange = <Label circular color='yellow' inverted><Icon name='time' /> 1 week</Label>;
    } else if (age > 7 && age <= 14) {
      statusChange = <Label circular color='yellow' inverted><Icon name='time' /> 1-2 weeks</Label>;
    } else if (age > 14 && age <= 21) {
      statusChange = <Label circular color='orange' inverted><Icon name='time' /> 2-3 weeks</Label>;
    } else if (age > 21) {
      statusChange = <Label circular color='red' inverted><Icon name='time' /> >3 weeks</Label>;
    }

    const googleLink = `https://calendar.google.com/calendar/render?action=TEMPLATE&text=${job.position}+Interview+at+${job.company.name}`;

    if (this.state.status === 'Interview Scheduled') {
      calendar = <Button target='_blank' href={googleLink} size='mini' floated='right' color='blue' circular icon='add to calendar'/>;
    } else {
      calendar = null;
    }

    var companyLink = `http://${job.company.company_url}`;

    return (
      <div>
        <Grid centered>
          <Grid.Row>
            <Card className='job-card'>
              <Card.Content >
                <Image target='_blank' href={companyLink} floated='left' size='tiny' src={job.company.logo_url} />
                <Card.Header>{job.company.name}</Card.Header>
                <Card.Meta>
                  {job.position}
                  <br/><br/>
                  {statusChange}
                  {calendar}
                  <ExpandedForm job={job} />
                  <EditForm job={job} x={x}/>
                </Card.Meta>
              </Card.Content>
            </Card>
          </Grid.Row>
        </Grid>
      </div>
    );
  }
}

export default JobCard;
