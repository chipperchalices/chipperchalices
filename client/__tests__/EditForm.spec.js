import React from 'react';
import expect from 'expect';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import JobCard from '../src/containers/JobCard.jsx';
import EditForm from '../src/containers/EditForm.jsx';

// import request from '__mocks__/request.js';
// import { Provider } from 'react-redux';
// import configureStore from 'redux-mock-store';
import fakeData from '../src/dummy_data.js';
import { lightBaseTheme, MuiThemeProvider, getMuiTheme } from 'material-ui/styles';


describe('Snapshot (EditForm)', () => {
  const renderer = new ShallowRenderer();
  it('EditForm should render using fakeData', () => {
    const wrap = renderer.render(
      <MuiThemeProvider>
        <EditForm job={fakeData[0]}/>
      </MuiThemeProvider>);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
  
  it('EditForm should render using fakeData (enzyme)', () => {
    const wrapper = shallow(
      <MuiThemeProvider>
        <EditForm job={fakeData[0]}/>
      </MuiThemeProvider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});

describe('Editform Props ', () => {
  let form;
  beforeEach(()=> {
    form = shallow(<MuiThemeProvider><EditForm job={fakeData[0]}/></MuiThemeProvider>);
  });
  
  it('should Receive All Props', () => {
    expect(Object.keys(form.props().job).length).toEqual(7);
  });

  it('should have position', () => {
    expect(form.props().job.position).toEqual('UI Developer');
  });

  it('should have position id', () => {
    expect(form.props().job.id).toEqual(1);
  });

  it('should have company name', () => {
    expect(form.props().job.company.name).toEqual('Vlocity');
  });

  it('should have company id', () => {
    expect(form.props().job.company.id).toEqual(12);
  });

  it('should have date', () => {
    expect(form.props().job.date).toEqual('2017-09-02');
  });

  it('should have application url', () => {
    expect(form.props().job.applicationUrl).toEqual('https://boards.greenhouse.io/vlocity/jobs/781103');
  });
  
});

// describe('Editform State Should Render Items', () => {
  
//   let form;
//   const muiTheme = getMuiTheme();
//   beforeEach(() => {
//     // mount = createMount();
//     form = mount(<MuiThemeProvider><EditForm job={fakeData[0]}/></MuiThemeProvider>);
//     form = form.find(EditForm);
//   });
//   // afterEach(() => {
//   //   mount.cleanUp();
//   // });

//   it('should call componentDidMount once', () => {
    
//     sinon.spy(EditForm.prototype, 'componentDidMount');
//     expect(EditForm.prototype.componentDidMount.calledOnce).toEqual(true);
//   });

//   it('should not be open when launched', () => {
//     expect(form.state().open).toEqual(false);
//   });
// });

describe('EditForm Update Button and Function', () => {
  
  let form;
  const muiTheme = getMuiTheme();
  beforeEach(() => {
    form = mount(<EditForm job={fakeData[0]}/>, {
      context: {muiTheme},
      childContextTypes: {muiTheme: React.PropTypes.object}
    });
  });

  it('should allow to open and close form', () => {
    expect(form.state().open).toEqual(false);
    form.find('FlatButton').find('button').simulate('click');
    expect(form.state().open).toEqual(true);
    form.find('FlatButton').find('button').simulate('click');
    expect(form.state().open).toEqual(false);
  });

  it('should not send anything if no change', () => {
    expect(toJson(form.state().job)).toEqual(toJson(fakeData));
    var result = form.instance().saveJob();
    expect(toJson(form.state().job)).toEqual(toJson(fakeData));
    expect(result).toEqual('No Changes');
  });
  
  //Can potentially refactor to use stub API mocking
  it('should send new items if change', () => {
    expect(toJson(form.state().job)).toEqual(toJson(fakeData));
    form.setState({company: 'Booga Booga', change: true});
    var result = form.instance().saveJob();
    expect(result).not.toEqual('No Changes');
  });
});

describe('Editform handle functions should change state', () => {

  let form;
  const muiTheme = getMuiTheme();

  beforeEach(() => {
    form = mount(<EditForm job={fakeData[0]}/>, {
      context: {muiTheme},
      childContextTypes: {muiTheme: React.PropTypes.object}
    });
  });

  it('should change state to be clicked', () => {
    expect(toJson(form.state())).toEqual(toJson(fakeData));
    expect(form.state().change).toEqual(false);
    
    form.instance().handleClick();
    
    expect(form.state().change).toEqual(false);
    expect(form.state().open).toEqual(true);
    expect(toJson(form.state())).toEqual(toJson(fakeData));
  });

  it('should change title', () => {
    expect(toJson(form.state())).toEqual(toJson(fakeData));
    expect(form.state().change).toEqual(false);
    expect(form.state().title).not.toEqual('Booga Booga');
    
    form.instance().handleTitle({target: {value: 'Booga Booga'}});
    
    expect(form.state().change).toEqual(true);
    expect(form.state().title).toEqual('Booga Booga');
  });

  //Add test for company id as well later
  it('should change company name', () => {
    expect(toJson(form.state())).toEqual(toJson(fakeData));
    expect(form.state().change).toEqual(false);
    expect(form.state().company).not.toEqual('Constantinople');
    
    form.instance().handleCompany({target: {value: 'Constantinople'}});
    
    expect(form.state().change).toEqual(true);
    expect(form.state().company).toEqual('Constantinople');
  });

  it('should change notes', () => {
    expect(toJson(form.state())).toEqual(toJson(fakeData));
    expect(form.state().change).toEqual(false);
    expect(form.state.notes).not.toEqual('What a test!');
    
    form.instance().handleNotes({target: {value: 'What a test!'}});
    
    expect(form.state().change).toEqual(true);
    expect(form.state().notes).toEqual('What a test!');
  });

  it('should change URL', () => {
    expect(toJson(form.state())).toEqual(toJson(fakeData));
    expect(form.state().change).toEqual(false);
    expect(form.state.url).not.toEqual('www.google.com/req/id/5000');
    
    form.instance().handleURL({target: {value: 'www.google.com/req/id/5000'}});
    
    expect(form.state().change).toEqual(true);
    expect(form.state().url).toEqual('www.google.com/req/id/5000');
  });

  it('should change Status', () => {
    expect(toJson(form.state())).toEqual(toJson(fakeData));
    expect(form.state().change).toEqual(false);
    expect(form.state().status).not.toEqual('Interviewed');
    expect(form.state().value).not.toEqual(3);

    form.instance().handleStatus({target: {textContent: 'Interviewed'}}, 3, 3);
    
    expect(form.state().change).toEqual(true);
    expect(form.state().status).toEqual('Interviewed');
    expect(form.state().value).toEqual(3);
  });

  it('should change Date (store as Date obj from string input)', () => {
    expect(toJson(form.state())).toEqual(toJson(fakeData));
    expect(form.state().change).toEqual(false);
    expect(form.state().date.toISOString().split('T')[0]).not.toEqual('2049-01-01');

    form.instance().handleDate({target: {value: 'BladeRunner 2049'}}, '2049-01-01');
  
    expect(form.state().change).toEqual(true);
    expect(form.state().date.toISOString().split('T')[0]).toEqual('2049-01-01');
  });

});