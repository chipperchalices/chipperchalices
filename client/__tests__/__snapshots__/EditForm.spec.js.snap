// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot (EditForm) EditForm should render using fakeData (enzyme) 1`] = `
<EditForm
  job={
    Object {
      "applicationUrl": "https://boards.greenhouse.io/vlocity/jobs/781103",
      "company": Object {
        "description": "Vlocity is a leading industry cloud software company that delivers award winning industry-specific applications on the Salesforce platform.",
        "id": 12,
        "location": "San Francisco, CA",
        "logoUrl": "https://tinyurl.com/ya54jrkq",
        "name": "Vlocity",
      },
      "currentStatus": "Interested",
      "date": "2017-09-02",
      "id": 1,
      "notes": null,
      "position": "UI Developer",
    }
  }
/>
`;

exports[`Snapshot (EditForm) EditForm should render using fakeData 1`] = `
<EditForm
  job={
    Object {
      "applicationUrl": "https://boards.greenhouse.io/vlocity/jobs/781103",
      "company": Object {
        "description": "Vlocity is a leading industry cloud software company that delivers award winning industry-specific applications on the Salesforce platform.",
        "id": 12,
        "location": "San Francisco, CA",
        "logoUrl": "https://tinyurl.com/ya54jrkq",
        "name": "Vlocity",
      },
      "currentStatus": "Interested",
      "date": "2017-09-02",
      "id": 1,
      "notes": null,
      "position": "UI Developer",
    }
  }
/>
`;
